PostCollection/script.js

```diff
@@ -0,0 +1,432 @@
+        },
+        {
+            id: '19',
+            description: 'super desc 19',
+            createdAt: new Date('2019-03-17T12:12:12'),
+            author: 'Agent 19',
+            likes: ['Peter', 'Arnold', 'Stalin'],
+            hashTags: ['USSR', 'Stalin']
+        },
+        {
+            id: '20',
+            description: 'super desc 20',
+            createdAt: new Date('2020-03-17T12:12:12'),
+            author: 'Agent 20',
+            likes: ['Peter', 'Arnold', 'Stalin'],
+            hashTags: ['USSR', 'Stalin']
+        }
+    ];
+
+    const requiredProperties = [
+        'id',
+        'description',
+        'createdAt',
+        'author'
+    ];
+
+    const optionalProperties = [
+        'likes',
+        'photoLinks',
+        'hashTags'
+    ];
+
+    const allProperties = [
+        'likes',
+        'photoLinks',
+        'hashTags',
+        'id',
+        'description',
+        'createdAt',
+        'author'
+    ];
+
+    class PostCollection {
+        _posts;
+        constructor(initialPosts) {
+            if (Array.isArray(initialPosts) && initialPosts.every(PostCollection._validatePost)) {
+                this._posts = initialPosts;
+            } else {
+                this._posts = undefined;
+            }
+        }
+
+        get(id) {
+            if (id && typeof id === 'string') {
+                return this._posts.find(item => item.id === id);
+            }
+            return undefined;
+        }
+
+        getPage(skip = 0, top = 10, filterConfig) {
+            this._posts.sort(PostCollection._compareDates);
+            PostCollection._validateFilterConfig(filterConfig);
+            return this._posts.filter(PostCollection._isValid, filterConfig);
+        }
+
+        edit(id, post) {
+            if (!PostCollection._validateProperties(post)) {
+                return false;
+            }
+            for (let i = 0; i < posts.length; ++i) {
+                if (this._posts[i].id === id) {
+                    let tmpKeys = Object.keys(post);
+                    for (let index = 0; index < tmpKeys.length; ++index) {
+                        let key =  tmpKeys[index];
+                        this._posts[i][key] = post[key];
+                    }
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        add(post) {
+            if (!PostCollection._validatePost(post)) {
+                return false;
+            }
+            this._posts.push(post);
+            return true;
+        }
+
+        remove(id) {
+            if (typeof id !== 'string') {
+                return false;
+            }
+
+            let removePostIndex = this._posts.findIndex(item => item.id === id);
+            if (removePostIndex === -1) {
+                return false;
+            }
+            this._posts.splice(removePostIndex, 1);
+            return true;
+        }
+
+        addAll(posts) {
```

Comment:
It can be done in a better way using `reduce` method, accumulating invalid posts